src/plaf/backends/halo2.rs

```diff
@@ -22,6 +22,44 @@ pub struct PlafH2Circuit {
     pub wit: Witness,
 }
 
+impl PlafH2Circuit {
+    pub fn instance<F: PrimeField<Repr = [u8; 32]>>(&self) -> Vec<Vec<F>> {
+        if !self.plaf.columns.public.is_empty() {
+            let mut instance_with_offsets = Vec::new();
+            for copy in &self.plaf.copys {
+                let (left_col, right_col) = &copy.columns;
+
+                let (witness_col, offsets) = match (left_col.kind, right_col.kind) {
+                    (ColumnKind::Witness, ColumnKind::Public) => (left_col, copy.offsets.clone()),
+                    (ColumnKind::Public, ColumnKind::Witness) => (
+                        right_col,
+                        copy.offsets.iter().map(|(l, r)| (*r, *l)).collect(),
+                    ),
+                    _ => continue,
+                };
+
+                for (witness_offset, public_offset) in offsets {
+                    instance_with_offsets.push((
+                        self.wit.witness[witness_col.index][witness_offset]
+                            .as_ref()
+                            .unwrap(),
+                        public_offset,
+                    ));
+                }
+            }
+            instance_with_offsets.sort_by_key(|&(_, offset)| offset);
```

Comment:
This seems a bit fragile because:
- If an offset is not used, the values after it will not appear at the correct offset
- If a public input is used in two copy constraints, it will appear twice with the same offset.

I would suggest following a different approach:
1. Create an empty vector called `instance`
2. In the loop at line 41 do the following:
  - if `instance.len() < public_offset+1` then `instance.resize(public_offset+1, F::zero())`
  - `instance[public_offset] = self.wit.witness[...].to_field()`

Reply:
Wonderful solution! Thanks!!